### PR TITLE
[SDK] Add a test covering exception thrown in custom sampler

### PR DESCRIPTION
**Pull request by CodeBlanch**
## Changes

* Adds a test covering what happens when an exception is thrown in a custom sampler. Intended to spark some discussion because the current behavior might be flawed?

/cc @alanwest @utpilla @cijothomas

### CODE REVIEW

**Comment by reyang** [comment]
A related question - what is the expected behavior if the sampler throws?

Here are some possible options:
* Exception from a sampler should be considered as `SamplingDecision.Drop`. In additional, an internal error log might be generated.
* A sampler should not throw exceptions, any exception from a sampler is considered as a bug of the sampler and the SDK will not handle it (which means the result is undefined and might lead to application crash).

**Comment by reyang** [comment]
I personally would vote for leaving it as undefined behavior, and not handle it in the SDK. This also applies to the following situations:
1. What if the sampler itself ended up creating more activities (which might cause recursive calls with stack overflow)?
2. What if the sampler is taking indefinite amount of time?
3. What if the sampler has contention that might interfere with the normal execution of the application?

File: test/OpenTelemetry.Tests/Trace/SamplersTest.cs [hunk 1]
```diff
@@ -252,5 +252,29 @@ public void SamplersDoesNotImpactTraceStateWhenUsingNull(SamplingDecision sampli
...
+                .Build();
+
+            Assert.Throws<InvalidOperationException>(() => activitySource.StartActivity("ThrowingSampler"));
+        }
+
+        private sealed class ThrowingSampler : Sampler
+        {
+            public override SamplingResult ShouldSample(in SamplingParameters samplingParameters)
+            {
+                throw new InvalidOperationException("ThrowingSampler");
```
Comment (utpilla) [inline test/OpenTelemetry.Tests/Trace/SamplersTest.cs:276]: I think the direction about who is responsible for catching the exception should come from the spec. Similar to how they have specified it for Processor `OnStart` and `OnEnd` methods [here](https://github.com/open-telemetry/opentelemetry-specification/blob/1e48fc558e7e259cccba097ca61880eae98d708b/specification/trace/sdk.md#onendspan).

If we have a particular preference, we should try to get the spec updated for [ShouldSample ](https://github.com/open-telemetry/opentelemetry-specification/blob/1e48fc558e7e259cccba097ca61880eae98d708b/specification/trace/sdk.md#shouldsample) first.

Comment (reyang) [reply test/OpenTelemetry.Tests/Trace/SamplersTest.cs:276]: > I think the direction about who is responsible for catching the exception should come from the spec. Similar to how they have specified it for Processor `OnStart` and `OnEnd` methods [here](https://github.com/open-telemetry/opentelemetry-specification/blob/1e48fc558e7e259cccba097ca61880eae98d708b/specification/trace/sdk.md#onendspan)

The spec doesn't seem to talk about who should be responsible if exception was thrown from a Processor's `OnStart`/`OnEnd`?

Comment (utpilla) [reply test/OpenTelemetry.Tests/Trace/SamplersTest.cs:276]: > OnEnd is called after a span is ended (i.e., the end timestamp is already set). This method MUST be called synchronously within the [Span.End() API](https://github.com/open-telemetry/opentelemetry-specification/blob/1e48fc558e7e259cccba097ca61880eae98d708b/specification/trace/api.md#end), therefore **it should not block or throw an exception**.

I interpret this statement as "Processor.OnEnd should not throw an exception" which makes it the user's responsibility to provide a spec-compliant processor to the SDK. If the processor logic can throw an exception, it's the user's responsibility to ensure that it is not thrown (they could simply wrap a try-catch around their processor logic.

**Comment by utpilla** [comment]
Another related question: Why do we not have a try-catch around `Exporter.Export` in BatchExportProcessor? https://github.com/open-telemetry/opentelemetry-dotnet/blob/f98f8fe166d59de008309e6e1ea3afaa5379afd5/src/OpenTelemetry/BatchExportProcessor.cs#L279-L296


We have try-catch around `Exporter.Export` for SimpleExporterProcessor and BaseExportingMetricReader:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/f98f8fe166d59de008309e6e1ea3afaa5379afd5/src/OpenTelemetry/SimpleExportProcessor.cs#L44-L54

https://github.com/open-telemetry/opentelemetry-dotnet/blob/f98f8fe166d59de008309e6e1ea3afaa5379afd5/src/OpenTelemetry/Metrics/BaseExportingMetricReader.cs#L90-L110

**Comment by cijothomas** [comment]
See https://github.com/open-telemetry/opentelemetry-dotnet/issues/1640

**Comment by CodeBlanch** [comment]
I was just poking around the spec. The docs for the different extension points have varying degrees of clarity, but the [error handling doc](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md#basic-error-handling-principles) seems crystal clear 😄 

>## Basic error handling principles
>OpenTelemetry implementations MUST NOT throw unhandled exceptions at run time.

**Comment by reyang** [comment]
> The spirit of this doc says to me the primary goal is to not interrupt the function of the code being instrumented a la we should try/catch user code.

What would you consider as "user code"? E.g. if someone is writing a processor/exporter, do we consider them as SDK developers or SDK plugin developers?

**Comment by CodeBlanch** [comment]
> What would you consider as "user code"? 

Let me amend my statement:

_The spirit of this doc says to me the primary goal is to not interrupt the function of the code being instrumented a la we should try/catch ~user~ all code._

To say that another way, what I feel like the spec is trying to say is: Once started, the SDK should never throw. Regardless of if it is itself throwing the exception or it is bubbling up an exception from some plugin it executed.

**Comment by reyang** [comment]
> To say that another way, what I feel like the spec is trying to say is: Once started, the SDK should never throw. Regardless of if it is itself throwing the exception or it is bubbling up an exception from some plugin it executed.

Putting these together, how would you phrase the implementation principles/guidelines? I think either way is fine as long as we have clear principles, and the worst situation is that we don't have principles + we never thought about it + we just do things randomly without clarity.

Take an example:

* https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#onstart "OnStart is called when a span is started. This method is called synchronously on the thread that started the span, therefore it should not block or throw exceptions."
* Do you expect the spec to be amended, with something like "OnStart ... should not block or throw exceptions. However, if OnStart throws exceptions, the SDK should swallow the exception and treat it as Behavior A, and follow up with Behavior B"

Here are more examples:
  * "If an exporter has thrown an exception to the caller, the exporter will be treated as faulty. As a result, the exception MUST be swallowed by the caller, followed by an internal error log emitted by the caller, and the caller should stop making any further call to the exporter"
  * "If a sampler has thrown an exception to the caller, the sampler will be treated as faulty. As a result, the `SamplingResult` should be treated as DROP + no extra attributes + bypass tracestate"

**Comment by CodeBlanch** [comment]
@reyang

My personal opinion, agree with @utpilla it would be best for the spec to prescribe the behavior. Ambiguity in specifications is bad. What should the spec say? IMO, keep it simple:

_OnStart ... should not block or throw exceptions. However, if OnStart throws exceptions, the SDK should swallow the exception and write the details to whatever diagnostic mechanism is in place for the SDK._

My reasoning is that the spec has been ambiguous and our code has been unprotected thus far, and the world hasn't ended, so we shouldn't over engineer something 😄

**Comment by reyang** [comment]
> @reyang
> 
> My personal opinion, agree with @utpilla it would be best for the spec to prescribe the behavior. Ambiguity in specifications is bad. What should the spec say? IMO, keep it simple:
> 
> _OnStart ... should not block or throw exceptions. However, if OnStart throws exceptions, the SDK should swallow the exception and write the details to whatever diagnostic mechanism is in place for the SDK._

I feel this simple answer is starting to scratch the surface of the design, which leads to couple more questions:
* Do we keep calling the sampler later (when new telemetry data is flowing through), assuming that it might stop throwing exceptions? - that might result in massive amount of exceptions and how do we handle that?
* Or we should somehow put the sampler in the abandoned list and never attempt to call it again?
* If we swallow the exception, what do we do with the telemetry - treat it as Drop or something else?

**Comment by CodeBlanch** [comment]
This is just my opinion but for the sampler I would say a throw is a drop. I would not prescribe back-off from the spec or anything more perverse. It could leave the door open for that should an SDK want it. My reasoning is a sampler can implement its own retry/backoff/fallback/whatever if it is doing complex things. The goal is just to fulfil the spec's mission to be transparent and not bring down the process/interfere with useful code the app is trying to execute.

**Comment by reyang** [comment]
I don't feel this would solve any problems or make things better, and it is more likely to make things worse. I'll take one example - a very bad sampler which triggers a blocking I/O operation:
1. There isn't much to do from the SDK, the user will for sure be impacted.
2. The sampler should not do such blocking call - the only reasonable fix is going to be provided by the developer who owns the sampler code, or a mitigation on the application developer to avoid using the sampler.

I consider taking blocking calls, making recursive calls, leaking memory, don't properly handle exceptions - these are equally bad, and from the consistency perspective it makes more sense to treat them in the same way.

More thinking regarding the ecosystem, completely agree with:

> My reasoning is that the spec has been ambiguous and our code has been unprotected thus far, and the world hasn't ended, so we shouldn't over engineer something

I think a simple "solution" is to do nothing (I guess LoggingProvider is doing the same - not over protecting), not over protecting things knowing that we won't be able to protect it and on the other side we might introduce bad behavior (e.g. a plugin developer might decide to throw exception from a deep callstack for "simplicity" instead of return SamplingDecision.Drop).

**Comment by cijothomas** [comment]
@alanwest good points.
Lets be good citizens by reporting bad actors to the cop, i.e if a bad actor throws, lets not eat it, and encourage bad behavior. instead, let the app crash. We are not violating Otel principles by doing so.

**Review by cijothomas — APPROVED** [review]

File: test/OpenTelemetry.Tests/Trace/SamplersTest.cs [hunk 2]
```diff
@@ -252,5 +252,29 @@ public void SamplersDoesNotImpactTraceStateWhenUsingNull(SamplingDecision sampli
                 Assert.Equal(parentTraceState, activity.TraceStateString);
             }
         }
+
+        [Fact]
+        public void SamplerExceptionBubblesUpTest()
+        {
+            // TODO: Discuss: An exception thrown in custom sampler probably
+            // should not blow up like this?
```
Comment (alanwest) [inline test/OpenTelemetry.Tests/Trace/SamplersTest.cs:260]: Remove TODO assuming we agree this is the behavior we want.

**Review by alanwest — APPROVED** [review]

**Comment by cijothomas** [comment]
It'd be good to modify the docs (separately) meant for custom plugin writers, to strongly warn them of the implications of throwing from their components and be explicit like "if you throw, it might lead to app crash"/similar wordings.
https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/trace/extending-the-sdk

File: test/OpenTelemetry.Tests/Trace/SamplersTest.cs (hunk 2 above)
Comment (cijothomas) [reply test/OpenTelemetry.Tests/Trace/SamplersTest.cs:260]: or leave a comment saying this is by design, and link to this PR/issues, so a future reader can easily find all the prior discussions on this topic.

Comment (CodeBlanch) [reply test/OpenTelemetry.Tests/Trace/SamplersTest.cs:260]: Updated

**Review by reyang — APPROVED** [review]

**Pull request merged**